fix indefinite retrying for `cockroach quit` when quorum is lost #14620

xphoniex opened the pull request:
error log is still misleading, should we call it a graceful shutdown when a hard shutdown has been initiated?

 

```
timed out, trying again
ok
initiating graceful shutdown of server
server drained and shutdown completed
```

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/14708)
<!-- Reviewable:end -->

a-robinson commented:
Thanks for the contribution, @xphoniex! Assigning @asubiotto for review since he's been looking into this as well, I believe.

tamird commented:
This doesn't look right to me - we want the node to stop, not for the client to give up.

xphoniex commented:
@tamird , it doesn't give up, it proceeds to initiate a hard shutdown once the one minute window has passed, I checked on my machine and all processes had been shutdown.

tamird commented:
Ah, my mistake. Looks like you have two commits here where there should be one?

asubiotto reviewed (commented):
Thanks a lot for the contribution! You can rebase the two commits into one using `git rebase -i HEAD~2` as outlined [here](http://stackoverflow.com/questions/12522565/how-can-i-combine-two-commits-into-one-commit).
Additionally, you can automatically close the issue this refers to by putting `Fixes #14620` on its own line in the commit message (run `git commit --amend`).
It would also be nice to provide a little more information of what is being done here in the commit message (e.g. "client initiates a hard shutdown on the server after a minute...").

asubiotto commented:
One thing I forgot to mention: `initiating graceful shutdown of server` etc... is printed because the  `stopper.ShouldStop()` case is being hit [here](https://github.com/cockroachdb/cockroach/blob/master/pkg/cli/start.go#L516) and we keep on going to print the messages on the server. This case simply waits for a signal that the stopper is stopped externally and I therefore think that it should be fine to `return nil` there since there is nothing to be done from the goroutine that prints the draining messages. @tamird what do you think?

xphoniex commented:
@asubiotto it should be okay now.

xphoniex commented:
Also, there's no guideline for writing the doc. Is this something I should do ? :)

tamird commented:
@xphoniex there's something not right with these commits.

xphoniex commented:
@tamird will fix it in a sec

asubiotto commented:
The changes look good. The TeamCity build is failing because of formatting issues in `pkg/cli/start.go`. You can run `gofmt -w pkg/cli/start.go` to fix this.

Note that "Fixes #14620" only works if it's on its own line. While you're modifying the commit message it would be good to fix the spelling of quorum and prepend the first line with the package name to know which package is affected by the commit as follows:

```
cli: fix indefinite retrying for `cockroach quit` when quorum is lost

Fixes #14620

add a timeout of 1 minute inside runQuit, after which a hard shutdown is initiated
```

asubiotto reviewed (approved):
LGTM

asubiotto commented:
Thanks for this PR @xphoniex

andreimatei commented:
@xphoniex You'll have to resolve conflicts with #14775. Sorry about that.

I think you shouldn't mark #14620 as resolved; we should keep it open and make a node be able to drain satisfactorily even when there's no quorum. Perhaps you can update the title of the issue to that when this PR merges.
I think in the future we'll probably want `cockroach quit` to be less liberal with brutal killing the process in the face of drain failures or timeouts. Probably depending on a flag, people should be able to depend on it to do proper draining, as draining becomes a tool for different kinds of machine or datacenter migrations.

---

Review status: 0 of 1 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14708#-:-KhTPcwiwFCPGrIYaTyF:bgl8jlu)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
> I think in the future we'll probably want cockroach quit to be less liberal with brutal killing the process in the face of drain failures or timeouts. Probably depending on a flag, people should be able to depend on it to do proper draining, as draining becomes a tool for different kinds of machine or datacenter migrations.

I disagree. A node that fails to die when you ask it to is more disruptive to a rollout process than one that fails to release all its leases (remember that graceful shutdown can never be 100% guaranteed because processes can always die for unrelated reasons). We should take care that this only happens in exceptional situations, but when those situations arise it's better to die abruptly than wait.

andreimatei commented:
Regardless of what we want `cockroach quit` to do, I think that one of the draining phases hanging indefinitely because the node was unable to write an updated liveness record is funky. Do we agree on this?

About the cli quit command, what I had in mind was the kind of draining that moves ranges away and waits for the up-replication to be done.

bdarnell commented:
> Regardless of what we want cockroach quit to do, I think that one of the draining phases hanging indefinitely because the node was unable to write an updated liveness record is funky. Do we agree on this?

Yes. But if it's happening because this is the last node being shut down and the cluster has lost quorum, there's nothing we can do but put a timeout on that liveness write. That's what I mean by it being better to die abruptly than wait indefinitely.

> About the cli quit command, what I had in mind was the kind of draining that moves ranges away and waits for the up-replication to be done.

We should have some way to do that, but not the default `cockroach quit`. The default assumption should be that the node is being taken down for something like a software update or reboot and will be coming back up with its data. In this case we want to relinquish all leases but keep the ranges so we can reuse them when we come back up.

xphoniex commented:
@andreimatei the conflict is when checkNodeRunning is returning error, right? want to be sure I'm not missing anything else

@andreimatei we could return after `<-stopper.IsStopped()` but we won't be saving shutdown logs. we can move the `msgDrain` inside the switch case (case sig) and we'd be only left with `server drained and shutdown completed` on console log

andreimatei commented:
I just meant that there's going to be a merge conflict; I wasn't trying to suggest something in particular to pay attention to.

I'm not sure what your second question is referring to :)

---

Review status: 0 of 1 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14708#-:-KhYeXZ6xelNXklHeeH0:bcw4j4d)*
<!-- Sent from Reviewable.io -->

xphoniex commented:
@andreimatei I just rebased my fork and it didn't give me a merge conflict :/

xphoniex committed:
cli: fix indefinite retrying for `cockroach quit` when quorum is lost

Fixes #14620

added a timeout of 1 minute inside runQuit, after which a hard shutdown is initiated

andreimatei commented:
Review status: 0 of 1 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/cli/start.go, line 516 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14708#-KhdbYM3zpOUETF7px6N:-KhdbYM3zpOUETF7px6O:b-nj74gz) ([raw file](https://github.com/cockroachdb/cockroach/blob/93cc469847d81d3b1466cc929aa90d73dbc469ce/pkg/cli/start.go#L516)):*
> ```Go
> 		return err
> 	case <-stopper.ShouldStop():
> 		<-stopper.IsStopped()
> ```

I'm confused about this new code. So before we were initiating draining on `stopped.ShouldStop`. Now, on `ShouldStop`, we seemingly block until `IsStopped` without doing the draining, and then we return directly, but we print a message about draining being done. What's the intention / what's going on?

---

*[pkg/cli/start.go, line 519 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14708#-KhdamgO-YOaVVdkIj_m:-KhdamgP7IxXjfuZjR-4:b-ngkjac) ([raw file](https://github.com/cockroachdb/cockroach/blob/93cc469847d81d3b1466cc929aa90d73dbc469ce/pkg/cli/start.go#L519)):*
> ```Go
> 		const msgDone = "server drained and shutdown completed"
> 		log.Infof(shutdownCtx, msgDone)
> 		fmt.Fprintln(os.Stdout, msgDone)
> ```

@knz how come no acceptance tests failed because of this unexpected new message? Is it surprising?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14708)*
<!-- Sent from Reviewable.io -->

asubiotto commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/cli/start.go, line 516 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14708#-KhdbYM3zpOUETF7px6N:-KhgraJW4PnEARCZQ7f9:b-wtdayg) ([raw file](https://github.com/cockroachdb/cockroach/blob/93cc469847d81d3b1466cc929aa90d73dbc469ce/pkg/cli/start.go#L516)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Andrei I think this just means that once IsStopped()'s channel unblocks, that means the server has finished draining.
But yes I also find it difficult to understand locally, it probably deserves a comment.
</blockquote></details>

This `ShouldStop` check is for any external requests to stop (for example the `quit` endpoint which hits `Drain` in `pkg/server/admin.go`). We previously would print `initiating graceful shutdown of server` and then wait for the stopper to be stopped after which we would print out `server drained and shutdown completed` even though we potentially got a request to perform a hard shutdown through the `quit` endpoint. I suggested to add this wait to avoid printing confusing log messages. I would remove outputting `msgDone` here @xphoniex and not print anything out (I think the responsibility of that should lie with the `quit` endpoint in this case), but add a comment as to why this wait is happening. Something to be worked on is to share more logic (and log messages) between these two endpoints but for now I think it's better to not output misleading log messages.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14708)*
<!-- Sent from Reviewable.io -->

xphoniex commented:
@andreimatei We were not initiating draining on `ShouldStop()` before. It was there for the Server to escape the select and log relevant messages which were confusing in this case, read post #1 & #8 here.

By the time we hit the `ShouldStop()` case, we've already drained the server. ( @asubiotto please correct me if I'm wrong )
